include/alpaka/core/Concepts.hpp

```diff
@@ -27,24 +27,62 @@ namespace alpaka
 
         namespace detail
         {
+            template<
+                typename TConcept,
+                typename TDerived>
+            struct ImplementsConcept {
+                template<
+                    typename TBase>
+                static auto implements(Implements<TConcept, TBase>&) -> std::true_type;
+                static auto implements(...) -> std::false_type;
+
+                using type = decltype(implements(std::declval<TDerived&>()));
```

Comment:
Shouldn't this part be `private`? Also, not sure if `type` as class name is consistent with the style (it may be, I'm just not sure)?

Reply:
Yes, I will make this private.

Reply:
I made all things that should not be accessible private. the `type` typedef ha been remove.

Reply:
After I made them private all compilers build correctly except gcc who complains:
```
Concepts.hpp:33:20: error: all member functions in class 'alpaka::concepts::detail::ImplementsConcept<TConcept, TDerived>' are private [-Werror=ctor-dtor-privacy]
```
The compiler is somehow right but the diagnostic is useless. I will remove the access-specifier because those classes are traits and within the `detail` namespace so there are two reasons not to instantiate them at all.

Reply:
Can you please approve this PR again?